tests/test_rpc.py

```diff
@@ -0,0 +1,412 @@
+        """
+        Store arg and return it.
+        """
+        self.arg = arg
+        return arg
+
+    def pass_named_arg(self, named_arg="arg"):
+        """
+        Store named_arg and return it.
+        """
+        self.named_arg = named_arg
+        return named_arg
+
+    def set_something(self, something):
+        """
+        Store something and return it.
+        """
+        self.something = something
+        return something
+
+    def boom(self):
+        """
+        This method raises an error which should be re-raised on the
+        other side of the RPC.
+        """
+        raise Boom()
+
+    def long_call(self, close_proxy, return_value=None):
+        """
+        Simulate a long call. Optionally close the proxy while this is
+        happening.
+
+        :returns: The value specified in return_value.
+        """
+        if close_proxy:
+            self.proxy.close()
+        time.sleep(3)
+        return return_value
+
+
+@pytest.fixture
+def fake_engine():
+    """
+    Fixture returning the Fake engine instance.
+    """
+    return FakeEngine()
+
+
+# We need to test the Http and Multiprocessing proxy connecting to each server
+# type.
+# Note that on Python 3, the multiprocessing server is never used by Desktop,
+# so we can ignore those servers.
+@pytest.fixture(
+    params=(
+        [
+            (DualRPCServer, MultiprocessingRPCProxy),
+            (MultiprocessingRPCServerThread, MultiprocessingRPCProxy),
+            (DualRPCServer, HttpRPCProxy),
+        ]
+        if six.PY2
+        else []
+    )
+    + [(HttpRPCServerThread, HttpRPCProxy)]
+)
+def server(fake_engine, request):
+    """
+    Readies an RPCServerThread for use, preconfigure to be able to
+    call method on a FakeEngine instance.
+
+    :returns: A RPCServerThread instance.
+    """
+    server_class = request.param[0]
+    client_class = request.param[1]
+
+    server = server_class(fake_engine)
+    server.register_function(fake_engine.pass_arg)
+    server.register_function(fake_engine.pass_named_arg)
+    server.register_function(fake_engine.boom)
+    server.register_function(fake_engine.long_call)
+    server.register_function(fake_engine.pass_arg, "pass_arg_as_another_name")
+    server.start()
+
+    if server_class == DualRPCServer:
+        if client_class == HttpRPCProxy:
+            factory = lambda server: client_class(server.pipes[1], server.authkey)
+        else:
+            factory = lambda server: client_class(server.pipes[0], server.authkey)
+    else:
+        factory = lambda server: client_class(server.pipe, server.authkey)
+
+    server.client_factory = factory
+    try:
+        yield server
+    finally:
+        if server.is_closed() is False:
+            server.close()
+        server.join()
+
+
+@pytest.fixture
+def proxy(server):
+    """
+    Creates a proxy that can communicate with the server fixture.
+
+    :returns: A proxy instance.
+    """
+    client = server.client_factory(server)
```

Comment:
I find this very confusing.
`server` is a method/fixture in the global scope, and that method creates and yields a variable called `server` which contains a `client_factory` callable that wants to be passed `server`. 😕
I think maybe I get it?

Reply:
I added a comment 😄